cognite/client/_api/datapoints.py

```diff
@@ -1,117 +1,969 @@
+                ...     ],
+                ...     granularity="1d",
+                ...     start=datetime(1990, 1, 1, tzinfo=timezone.utc),
+                ...     end=datetime(2020, 12, 31, tzinfo=timezone.utc),
+                ...     uniform_index=True)
+
+            Get a pandas dataframe containing the 'average' aggregate for two time series using a 30 day granularity,
+            starting Jan 1, 1970 all the way up to present, without having the aggregate name in the column names::
+
+                >>> df = client.time_series.data.retrieve_dataframe(
+                ...     external_id=["foo", "bar"],
+                ...     aggregates=["average"],
+                ...     granularity="30d",
+                ...     include_aggregate_name=False)
+        """
+        _, pd = local_import("numpy", "pandas")  # Verify that deps are available or raise CogniteImportError
+        if column_names not in {"id", "external_id"}:
+            raise ValueError(f"Given parameter {column_names=} must be one of 'id' or 'external_id'")
+
+        query = DatapointsQuery(
+            start=start,
+            end=end,
+            id=id,
+            external_id=external_id,
+            aggregates=aggregates,
+            granularity=granularity,
             limit=limit,
+            include_outside_points=include_outside_points,
             ignore_unknown_ids=ignore_unknown_ids,
         )
-        dps_list = fetcher.fetch(query)
-        if is_single_id:
-            if len(dps_list) == 0 and ignore_unknown_ids is True:
-                return None
-            return dps_list[0]
-        return dps_list
+        fetcher = dps_fetch_selector(self, user_queries=[query])
+        if uniform_index:
+            grans_given = set(q.granularity for q in fetcher.all_queries)
+            is_limited = any(q.limit is not None for q in fetcher.all_queries)
+            if fetcher.raw_queries or len(grans_given) > 1 or is_limited:
+                raise ValueError(
+                    "Cannot return a uniform index when asking for aggregates with multiple granularities "
+                    f"({grans_given}) OR when (partly) querying raw datapoints OR when a finite limit is used."
+                )
+        df = fetcher.fetch_all_datapoints(use_numpy=True).to_pandas(column_names, include_aggregate_name)
+        if not uniform_index:
+            return df
+
+        start = pd.Timestamp(min(q.start for q in fetcher.agg_queries), unit="ms")
+        end = pd.Timestamp(max(q.end for q in fetcher.agg_queries), unit="ms")
+        (granularity,) = grans_given
+        # Pandas understand "Cognite granularities" except `m` (minutes) which we must translate:
+        return df.reindex(pd.date_range(start=start, end=end, freq=granularity.replace("m", "T"), inclusive="left"))
+
+    @overload
+    def query(
+        self,
+        query: Union[Sequence[DatapointsQuery], DatapointsQuery],
+        use_numpy: Literal[False],
```

Comment:
Same note as above, if we give DatapointsList and DatapointsArrayList a common Protocol, we won't need this argument. We just enable behind the scenes.

Reply:
I’ve done some more thinking and have come to the realization that I’m just gonna remove the query method entirely and allow individual customization of the ignore option. It will not mirror the API endpoint anymore, but imo be an improvement ☺️